proposals/gc/MVP.md

```diff
@@ -594,112 +594,52 @@ In particular, `ref.null` is typed as before, despite the introduction of `none`
   - note: this succeeds for all values; moreover, composing this with `extern.internalize` (in either order) yields the original value
 
 
-#### Classification
-
-* `ref.is_data` checks whether a reference is compound data
-  - `ref.is_data : [anyref] -> [i32]`
-
-* `ref.is_array` checks whether a reference is an array
-  - `ref.is_array : [anyref] -> [i32]`
-
-* `ref.is_i31` checks whether a reference is an i31
-  - `ref.is_i31 : [anyref] -> [i32]`
-
-* `br_on_data <labelidx>` branches if a reference is compound data
-  - `br_on_data $l : [t0* t] -> [t0* t]`
-    - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `(ref data) <: t'`
-  - passes operand along with branch as data, plus possible extra args
+#### Casts
 
-* `br_on_non_data <labelidx>` branches if a reference is not compound data
-  - `br_on_non_data $l : [t0* t] -> [t0* (ref data)]`
-    - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `t <: t'`
-  - passes operand along with branch, plus possible extra args
+Casts work for both abstract and concrete types. In the latter case, they test of the operands RTT is a sub-RTT of the target type.
 
-* `br_on_array <labelidx>` branches if a reference is an array
-  - `br_on_array $l : [t0* t] -> [t0* t]`
-    - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `(ref array) <: t'`
-  - passes operand along with branch as data, plus possible extra args
+* `ref.test null? <heaptype>` checks whether a reference is a given heap type
+  - `ref.test null? ht : [(ref null ht')] -> [i32]`
+    - iff `ht <: tht` and `ht' <: tht` where `tht` is a common super type
+  - if `null?` is present, returns 1 for null, otherwise 0
 
-* `br_on_non_array <labelidx>` branches if a reference is not an array
-  - `br_on_non_array $l : [t0* t] -> [t0* (ref array)]`
-    - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `t <: t'`
-  - passes operand along with branch, plus possible extra args
+* `ref.cast null? <heaptype>` tries to convert to a given heap type
+  - `ref.cast null? ht : [(ref null ht')] -> [(ref null2? ht)]`
+    - iff `ht <: tht` and `ht' <: tht` where `tht` is a common super type
```

Comment:
Just for my own understanding, is `ht <: ht'` an equivalent formulation of this condition, or is there some subtlety around subsumption that I don't have right in my head?